esrally/driver/driver.py

```diff
@@ -692,21 +693,38 @@ def calculate_worker_assignments(host_configs, client_count):
              in that list contains another list with the clients that should be assigned to these workers.
     """
     assignments = []
-
+    client_idx = 0
+    host_count = len(host_configs)
+    clients_per_host = math.ceil(client_count / host_count)
+    remaining_clients = client_count
     for host_config in host_configs:
-        assignments.append({
+        # the last host might not need to simulate as many clients as the rest of the hosts as we eagerly
+        # assign clients to hosts.
+        clients_on_this_host = min(clients_per_host, remaining_clients)
+        assignment = {
             "host": host_config["host"],
-            "workers": [[] for _ in range(host_config["cores"])],
-            "worker": 0
-        })
+            "workers": [],
+        }
+        assignments.append(assignment)
+
+        workers_on_this_host = host_config["cores"]
+        clients_per_worker = [0] * workers_on_this_host
+
+        # determine how many clients each worker should simulate
+        for c in range(clients_on_this_host):
+            clients_per_worker[c % workers_on_this_host] += 1
+
+        # assign client ids to workers
+        for worker_idx, client_count in enumerate(clients_per_worker):
+            worker_assignment = []
+            assignment["workers"].append(worker_assignment)
+            for c in range(client_idx, client_idx + client_count):
+                worker_assignment.append(c)
+            client_idx += client_count
+
+        remaining_clients -= clients_on_this_host
 
-    for client_idx in range(client_count):
-        host = assignments[client_idx % len(assignments)]
-        workers = host["workers"]
-        worker_idx = host["worker"]
-        worker = host["workers"][worker_idx % len(workers)]
-        worker.append(client_idx)
-        host["worker"] = worker_idx + 1
+    assert remaining_clients == 0
```

Comment:
My intention was that this is a lightweight check for violation of an invariant due a logical error in the implementation (in the sense of "this thing should never fail"). Should that ever fail it would be easy to reproduce and that's why I used an assertion.

Reply:
Ok I see used as a debugging aid then ++